feat: add `math/base/special/bernoullif`

gururaj1512 opened the pull request:
Progresses #649

## Description

> What is the purpose of this pull request?
- adds `math/base/special/bernoullif`, which would be the single-precision equivalent for [math/base/special/bernoullif](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/bernoulli).

```c
float stdlib_base_bernoullif( const int32_t n )
```

## Related Issues

> Does this pull request have any related issues?

This pull request:

- Progresses: #649 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

gururaj1512 committed:
feat:add math/base/special/bernoullif

kgryte commented on lib/node_modules/@stdlib/math/base/special/bernoullif/lib/main.js:
```diff
@@ -0,0 +1,102 @@
+	if ( isnanf( n ) || !isNonNegativeInteger( n ) ) {
+		return NaN;
+	}
+	n = float64ToFloat32( n );
```
I don't believe this is necessary. By way of L86, you already know that `n` is a nonnegative integer, and L93 checks for the maximum Bernoulli number.

gururaj1512 replied:
L86-L88 check is for `NaN` and `Negative` values & L93-L95 checks for input values whose bernoulli number exceeds `bernoullif` function's max limit and thereafter returns -INF & +INF.

gururaj1512 replied:
Both checks are necessary

kgryte replied:
It's not the checks which are unnecessary, it is the conversion from `float64` to `float32`. L89 is unnecessary.

kgryte commented on lib/node_modules/@stdlib/math/base/special/bernoullif/lib/bernoullif.json:
```diff
@@ -0,0 +1,34 @@
+[
+  1.00000000000000000000000000000000000000000,
```
These values all have very high precision. Is there a reason for that?

gururaj1512 replied:
No, I have used similar values as of `math/base/special/bernoulli`. Should I make it less precise?

gunjjoshi replied:
> No, I have used similar values as of math/base/special/bernoulli. Should I make it less precise?

Yes, @gururaj1512.

You'll have to keep these values in appropriate precision. You can use https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/number/float64/base/to-float32 for the conversion.

kgryte commented on lib/node_modules/@stdlib/math/base/special/bernoullif/src/main.c:
```diff
@@ -0,0 +1,85 @@
+};
+
+
+int32_t MAX_BERNOULLI = 62;
```
```suggestion
static const int32_t MAX_BERNOULLI = 62;
```

This variable should be local to the translation unit.

kgryte commented on lib/node_modules/@stdlib/math/base/special/bernoullif/manifest.json:
```diff
@@ -0,0 +1,84 @@
+      "libpath": [],
+      "dependencies": [
+        "@stdlib/math/base/napi/unary",
+        "stdlib/math/base/assert/is_nonnegative_integer.h",
+        "stdlib/math/base/assert/is_nanf.h",
+        "stdlib/math/base/assert/is_odd.h",
+        "stdlib/constants/float32/ninf.h",
+        "stdlib/constants/float32/pinf.h"
```
These paths are all incorrect and are the reason the tests are failing. See other packages and L39 for how to list dependencies.

gururaj1512 committed:
Update lib/node_modules/@stdlib/math/base/special/bernoullif/src/main.c

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Gururaj Gurram <143020143+gururaj1512@users.noreply.github.com>

gururaj1512 committed:
fix: manifest.json file dependencies

kgryte commented on lib/node_modules/@stdlib/math/base/special/bernoullif/manifest.json:
```diff
@@ -37,11 +37,11 @@
-        "stdlib/math/base/assert/is_odd.h",
-        "stdlib/constants/float32/ninf.h",
-        "stdlib/constants/float32/pinf.h"
+        "@stdlib/math/base/assert/is_nonnegative_integer.h",
```
These deps are still incorrect. You need to list package names, not paths to header files.

gururaj1512 committed:
fix: apply review changes

gururaj1512 committed:
fix: changes

gururaj1512 committed:
fix: check errors

gururaj1512 committed:
fix: bernoullif manifest.json file

Planeshifter commented on lib/node_modules/@stdlib/math/base/special/bernoullif/README.md:
```diff
@@ -0,0 +1,232 @@
+
+# bernoullif
+
+> Compute the nth [Bernoulli number][bernoulli-number] of a  single-precision floating-point number.
```
```suggestion
> Compute the nth [Bernoulli number][bernoulli-number] of a single-precision floating-point number.
```

gunjjoshi replied:
@gururaj1512 We can update this to:
```bash
Compute the nth [Bernoulli number][bernoulli-number] as a single-precision floating-point number.
```
here and elsewhere in this package, as we are taking an integer, not a floating-point number.

Planeshifter commented:
@gunjjoshi This is not a problem exclusively with this PR, since it also applies to `math/base/special/bernoulli`, which you added in https://github.com/stdlib-js/stdlib/pull/1860, but I noticed that the implementation returns zero for all odd numbers. However, Wikipedia says that this is incorrect for input value `1`: https://en.wikipedia.org/wiki/Bernoulli_number 
Could you please take a look and confirm whether we need to handle this case separately?

gunjjoshi commented:
@Planeshifter Thanks for catching this. After digging for some time, I found that there is a mix of opinions on whether `B(1) = +1/2` or `B(1) = -1/2`. For instance, here is a discussion from `sympy`: https://github.com/sympy/sympy/issues/23866.
While it is incorrect to use `B(1) = 0`, the choice that must be made is between `+1/2` or `-1/2`. 
`sympy` now uses `+1/2` (it used `-1/2` earlier). There's a discussion at `sagemath` too: https://github.com/sagemath/sage/issues/34521.

Overall, I think, it would be better if we too, use `B(1) = +1/2`. I'll make this change in [`math/base/special/bernoulli`](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/math/base/special/bernoulli), if this is fine.

gunjjoshi commented:
But yes, NIST prescribes the convention to use `B(1) = -1/2`: https://en.wikipedia.org/wiki/Bernoulli_number#Notation
What should we follow, in this case?

kgryte commented:
Let's go ahead and use 1/2. This will match SymPy and is also used as a convention in R. As commented elsewhere, how much the choice actually matters in practice is subject to some debate. In this case, I'm more inclined to go with Knuth.

kgryte commented:
Similar to the changes proposed in https://github.com/stdlib-js/stdlib/pull/3108, we'll want to make similar changes here.

gururaj1512 committed:
refactor: update return value for n=1

gururaj1512 committed:
fix: lint error

gururaj1512 commented:
Refactored it as suggested.

kgryte commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into bernoullif

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/bernoullif/src/main.c:
```diff
@@ -0,0 +1,86 @@
+	-2.84987693024508822262691464329106781609195e30,
+	2.38654274996836276446459819192192149717514e32,
+	-2.13999492572253336658107447651910973926742e34,
+	2.05009757234780975699217330956723102516667e36,
```
@gururaj1512 The next number here can also be represent as a single-precision floating-point number, right? We can include one more number here, i.e., `-2.0938006042234345e+38f`.

Also, you'll need to change all these numbers as single-precision floating point-numbers. For instance, the next number as double-precision floating-point number, in `bernoulli` is `-2.09380059113463784090951852900279701847092e38,`. But, in single-precision, we can use `-2.0938006042234345e+38f`.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/bernoullif/src/main.c:
```diff
@@ -0,0 +1,86 @@
+};
+
+
+static const int32_t MAX_BERNOULLI = 62;
```
Consequently, this would then be changed to `64` here and everywhere else in this package.

gunjjoshi commented:
@gururaj1512 You can represent numbers till `B(64)` in single-precision, rather than the current `B(62)` in this PR. You'll have to modify everything accordingly, in all files.

gururaj1512 committed:
refactor: as per suggested

gururaj1512 committed:
refactor: max_bernoullif to 64

kgryte commented:
@gururaj1512 Looks like this PR still has failing tests. We'll need to get those resolved before further review and subsequent merge.

gururaj1512 committed:
Update main.c

Signed-off-by: Gururaj Gurram <143020143+gururaj1512@users.noreply.github.com>

gururaj1512 commented:
> @gururaj1512 Looks like this PR still has failing tests. We'll need to get those resolved before further review and subsequent merge.

Resolved.

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/bernoullif/benchmark/benchmark.js:
@gururaj1512 We should pre-compute the values outside the loop, here and in `benchmark.native.js`, similar to what we do here: https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/bernoulli/benchmark/benchmark.js

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/bernoullif/benchmark/c/native/benchmark.c:
```diff
@@ -0,0 +1,133 @@
+*
+* @return elapsed time in seconds
+*/
+static double benchmark( void ) {
+	double elapsed;
+	double t;
+	float x;
+	float y;
+	int i;
+
+	t = tic();
+	for ( i = 0; i < ITERATIONS; i++ ) {
+		x = ( 500.0 * rand_float() );
+		y = stdlib_base_bernoullif( x );
+		if ( y != y ) {
+			printf( "should not return NaN\n" );
+			break;
+		}
+	}
+	elapsed = tic() - t;
+	if ( y != y ) {
+		printf( "should not return NaN\n" );
+	}
+	return elapsed;
+}
```
Similarly, we can declare `x` as an array and generate the values beforehand here too. See: https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/math/base/special/bernoulli/benchmark/c/native/benchmark.c

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/bernoullif/docs/repl.txt:
```diff
@@ -0,0 +1,45 @@
+
+{{alias}}( n )
+    Compute the nth Bernoulli number as a single-precision floating-point
```
```suggestion
    Computes the nth Bernoulli number as a single-precision floating-point
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/bernoullif/docs/types/index.d.ts:
```diff
@@ -0,0 +1,72 @@
+// TypeScript Version: 4.1
+
+/**
+* Compute the nth Bernoulli number as a single-precision floating-point number.
```
```suggestion
* Computes the nth Bernoulli number as a single-precision floating-point number.
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/bernoullif/include/stdlib/math/base/special/bernoullif.h:
```diff
@@ -0,0 +1,40 @@
+#endif
+
+/**
+* Computes the nth Bernoulli number.
```
```suggestion
* Computes the nth Bernoulli number as a single-precision floating-point number.
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/bernoullif/lib/main.js:
```diff
@@ -0,0 +1,104 @@
+// MAIN //
+
+/**
+* Compute the nth Bernoulli number as a single-precision floating-point number.
```
```suggestion
* Computes the nth Bernoulli number as a single-precision floating-point number.
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/bernoullif/lib/main.js:
```diff
@@ -0,0 +1,104 @@
+		return 0.0;
+	}
+	if ( n > MAX_BERNOULLI ) {
+		return ( ( n/2 ) & 1 ) ? PINF : NINF;
```
```suggestion
		return ( ( n / 2 ) & 1 ) ? PINF : NINF;
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/bernoullif/lib/main.js:
```diff
@@ -0,0 +1,104 @@
+	if ( n > MAX_BERNOULLI ) {
+		return ( ( n/2 ) & 1 ) ? PINF : NINF;
+	}
+	return float64ToFloat32( BERNOULLIF[ n/2 ] );
```
```suggestion
	return float64ToFloat32( BERNOULLIF[ n / 2 ] );
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/bernoullif/lib/native.js:
```diff
@@ -0,0 +1,77 @@
+// MAIN //
+
+/**
+* Compute the nth Bernoulli number as a single-precision floating-point number.
```
```suggestion
* Computes the nth Bernoulli number as a single-precision floating-point number.
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/bernoullif/src/main.c:
```diff
@@ -0,0 +1,87 @@
+static const int32_t MAX_BERNOULLI = 64;
+
+/**
+* Compute the nth Bernoulli number as a single-precision floating-point number.
```
```suggestion
* Computes the nth Bernoulli number as a single-precision floating-point number.
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/bernoullif/test/test.js:
```diff
@@ -0,0 +1,109 @@
+	var v;
+	var i;
+
+	t.strictEqual( isnanf( bernoullif( -3.14 ) ), true, 'returns NaN' );
```
```suggestion
	t.strictEqual( isnanf( bernoullif( -3.14 ) ), true, 'returns expected value' );
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/bernoullif/test/test.native.js:
```diff
@@ -0,0 +1,106 @@
+
+	for ( i = -1; i > -50; i-- ) {
+		v = bernoullif( i );
+		t.strictEqual( isnanf( v ), true, 'returns NaN when provided ' + i );
```
```suggestion
		t.strictEqual( isnanf( v ), true, 'returns expected value when provided ' + i );
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/bernoullif/test/test.native.js:
```diff
@@ -0,0 +1,106 @@
+	var v;
+	var i;
+
+	t.strictEqual( isnanf( bernoullif( -3.14 ) ), true, 'returns NaN' );
```
```suggestion
	t.strictEqual( isnanf( bernoullif( -3.14 ) ), true, 'returns expected value' );
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/bernoullif/test/test.native.js:
```diff
@@ -0,0 +1,106 @@
+	var i;
+	for ( i = 0; i < 66; i += 2 ) {
+		v = bernoullif( i );
+		t.strictEqual( v, float64ToFloat32( BERNOULLIF[ i/2 ] ), 'returns the '+i+'th Bernoulli number' );
```
```suggestion
		t.strictEqual( v, float64ToFloat32( BERNOULLIF[ i / 2 ] ), 'returns the '+i+'th Bernoulli number' );
```

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/bernoullif/test/test.js:
```diff
@@ -0,0 +1,109 @@
+	var i;
+	for ( i = 0; i < 66; i += 2 ) {
+		v = bernoullif( i );
+		t.strictEqual( v, float64ToFloat32( BERNOULLIF[ i/2 ] ), 'returns the '+i+'th Bernoulli number' );
```
```suggestion
		t.strictEqual( v, float64ToFloat32( BERNOULLIF[ i / 2 ] ), 'returns the '+i+'th Bernoulli number' );
```

gunjjoshi committed:
Apply suggestions from code review

Signed-off-by: Gunj Joshi <gunjjoshi8372@gmail.com>

gunjjoshi committed:
Update lib/node_modules/@stdlib/math/base/special/bernoullif/test/test.native.js

Signed-off-by: Gunj Joshi <gunjjoshi8372@gmail.com>

gururaj1512 committed:
refactor: benchmark files as per project conventions

gunjjoshi commented on lib/node_modules/@stdlib/math/base/special/bernoullif/README.md:
```diff
@@ -0,0 +1,232 @@
+
+#### bernoullif( n )
+
+Compute the nth [Bernoulli number][bernoulli-number] as a single-precision floating-point number.
```
```suggestion
Computes the nth [Bernoulli number][bernoulli-number] as a single-precision floating-point number.
```

gunjjoshi committed:
Update lib/node_modules/@stdlib/math/base/special/bernoullif/README.md

Signed-off-by: Gunj Joshi <gunjjoshi8372@gmail.com>

gunjjoshi commented:
/stdlib update-copyright-years

stdlib-bot committed:
chore: update copyright years

anandkaranubc commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into bernoullif

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/bernoullif/src/main.c:
```diff
@@ -0,0 +1,87 @@
+#include "stdlib/constants/float32/ninf.h"
+#include "stdlib/constants/float32/pinf.h"
+
+static const float bernoulli_value[ 33 ] = {
+	1.0000000000000000f,
+	0.1666666666666666f,
+	-0.0333333333333333f,
+	0.0238095238095238f,
+	-0.0333333333333333f,
+	0.0757575757575757f,
+	-0.2531135531135531f,
+	1.1666666666666666f,
+	-7.0921568627450980f,
+	54.9711779448621553f,
+	-529.1242424242424242f,
+	6192.1231884057971014f,
+	-86580.2531135531135531f,
+	1.4255171666666666e+6f,
+	-2.7298231067816091e+7f,
+	6.0158087390064236e+8f,
+	-1.5116315767092156e+10f,
+	4.2961464306116666e+11f,
+	-1.3711655205088332e+13f,
+	4.8833231897359316e+14f,
+	-1.9296579341940068e+16f,
+	8.4169304757368261e+17f,
+	-4.0338071854059455e+19f,
+	2.1150748638081991e+21f,
+	-1.2086626522296525e+23f,
+	7.5008667460769643e+24f,
+	-5.0387781014810689e+26f,
+	3.6528776484818123e+28f,
+	-2.8498769302450882e+30f,
+	2.3865427499683627e+32f,
+	-2.1399949257225333e+34f,
+	2.0500975723478097e+36f,
+	-2.0938006042234345e+38f
+};
```
```suggestion
static const float bernoulli_value[ 33 ] = {
	1.0f,
	0.1666666716337204f,
	-0.03333333507180214f,
	0.02380952425301075f,
	-0.03333333507180214f,
	0.07575757801532745f,
	-0.2531135678291321f,
	1.1666666269302368f,
	-7.092156887054443f,
	54.97117614746094f,
	-529.124267578125f,
	6192.123046875f,
	-86580.25f,
	1425517.125f,
	-27298232.0f,
	601580864.0f,
	-15116315648.0f,
	429614628864.0f,
	-13711654780928.0f,
	488332312182784.0f,
	-19296579391324160.0f,
	841693056053805000.0f,
	-40338073359287845000.0f,
	2.1150748918604188e+21f,
	-1.2086626472668042e+23f,
	7.500866956719485e+24f,
	-5.038777949942763e+26f,
	3.652877742276384e+28f,
	-2.849876996904201e+30f,
	2.386542781638813e+32f,
	-2.1399950070596233e+34f,
	2.050097633557977e+36f,
	-2.0938006042234345e+38f
};
```

anandkaranubc replied:
All these values were fixed using  https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/number/float64/base/to-float32 for the conversion.

kgryte replied:
@anandkaranubc Your change is fine, but also note that C would have automatically done the conversions for you. That said, it is less confusing if the displayed values actually match the runtime values so this change is a +0.5.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/bernoullif/src/main.c:
```diff
@@ -0,0 +1,87 @@
+	2.0500975723478097e+36f,
+	-2.0938006042234345e+38f
+};
+
+
+static const int32_t MAX_BERNOULLI = 64;
```
```suggestion

static const int32_t MAX_BERNOULLI = 64;
```

anandkaranubc replied:
Also, shouldn't it be `66`?

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/bernoullif/src/main.c:
```diff
@@ -0,0 +1,87 @@
+*
+* @example
+* float out = stdlib_base_bernoullif( 0 );
+* // returns 1
```
```suggestion
* // returns 1.0f
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/bernoullif/src/main.c:
```diff
@@ -0,0 +1,87 @@
+#include "stdlib/math/base/special/bernoullif.h"
+#include "stdlib/math/base/assert/is_odd.h"
+#include "stdlib/constants/float32/ninf.h"
+#include "stdlib/constants/float32/pinf.h"
```
```suggestion
#include <stdint.h>
```

Because you are using `int32_t`

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/bernoullif/src/main.c:
```diff
@@ -0,0 +1,87 @@
+*/
+float stdlib_base_bernoullif( const int32_t n ) {
+	if ( n < 0 ) {
+		return 0.0 / 0.0; // NaN
```
```suggestion
		return 0.0f / 0.0f; // NaN
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/bernoullif/src/main.c:
```diff
@@ -0,0 +1,87 @@
+		return 0.0 / 0.0; // NaN
+	}
+	if ( n == 1 ) {
+		return 0.5;
```
```suggestion
		return 0.5f;
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/bernoullif/src/main.c:
```diff
@@ -0,0 +1,87 @@
+		return 0.5;
+	}
+	if ( stdlib_base_is_odd( n ) ) {
+		return 0.0;
```
```suggestion
		return 0.0f;
```

anandkaranubc replied:
```suggestion
	if ( n & 1 ) {
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/bernoullif/src/main.c:
```diff
@@ -0,0 +1,87 @@
+*/
+
+#include "stdlib/math/base/special/bernoullif.h"
+#include "stdlib/math/base/assert/is_odd.h"
```
```suggestion
```

Not needed because of above.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/bernoullif/src/addon.c:
```diff
@@ -0,0 +1,23 @@
+#include "stdlib/math/base/special/bernoullif.h"
+#include "stdlib/math/base/napi/unary.h"
+
+// cppcheck-suppress shadowFunction
```
```suggestion
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/bernoullif/lib/native.js:
```diff
@@ -0,0 +1,77 @@
+
+/**
+* Computes the nth Bernoulli number as a single-precision floating-point number.
+*
```
```suggestion
* @private
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/bernoullif/lib/bernoullif.json:
```diff
@@ -0,0 +1,35 @@
+[
+  1.00000000000000000000000000000000000000000,
+  0.166666666666666666666666666666666666666667,
+  -0.0333333333333333333333333333333333333333333,
+  0.0238095238095238095238095238095238095238095,
+  -0.0333333333333333333333333333333333333333333,
+  0.0757575757575757575757575757575757575757576,
+  -0.253113553113553113553113553113553113553114,
+  1.16666666666666666666666666666666666666667,
+  -7.09215686274509803921568627450980392156863,
+  54.9711779448621553884711779448621553884712,
+  -529.124242424242424242424242424242424242424,
+  6192.12318840579710144927536231884057971014,
+  -86580.2531135531135531135531135531135531136,
+  1.42551716666666666666666666666666666666667e6,
+  -2.72982310678160919540229885057471264367816e7,
+  6.01580873900642368384303868174835916771401e8,
+  -1.51163157670921568627450980392156862745098e10,
+  4.29614643061166666666666666666666666666667e11,
+  -1.37116552050883327721590879485616327721591e13,
+  4.88332318973593166666666666666666666666667e14,
+  -1.92965793419400681486326681448632668144863e16,
+  8.41693047573682615000553709856035437430786e17,
+  -4.03380718540594554130768115942028985507246e19,
+  2.11507486380819916056014539007092198581560e21,
+  -1.20866265222965259346027311937082525317819e23,
+  7.50086674607696436685572007575757575757576e24,
+  -5.03877810148106891413789303052201257861635e26,
+  3.65287764848181233351104308429711779448622e28,
+  -2.84987693024508822262691464329106781609195e30,
+  2.38654274996836276446459819192192149717514e32,
+  -2.13999492572253336658107447651910973926742e34,
+  2.05009757234780975699217330956723102516667e36,
+  -2.09380059113463784090951852900279701847092e38
+]
```
```suggestion
[
  1.0,
  0.1666666716337204,
  -0.03333333507180214,
  0.02380952425301075,
  -0.03333333507180214,
  0.07575757801532745,
  -0.2531135678291321,
  1.1666666269302368,
  -7.092156887054443,
  54.97117614746094,
  -529.124267578125,
  6192.123046875,
  -86580.25,
  1425517.125,
  -27298232.0,
  601580864.0,
  -15116315648.0,
  429614628864.0,
  -13711654780928.0,
  488332312182784.0,
  -19296579391324160.0,
  841693056053805000.0,
  -40338073359287845000.0,
  2.1150748918604188e+21,
  -1.2086626472668042e+23,
  7.500866956719485e+24,
  -5.038777949942763e+26,
  3.652877742276384e+28,
  -2.849876996904201e+30,
  2.386542781638813e+32,
  -2.1399950070596233e+34,
  2.050097633557977e+36,
  -2.0938006042234345e+38
]
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/bernoullif/lib/main.js:
```diff
@@ -0,0 +1,104 @@
+
+// MODULES //
+
+var isNonNegativeInteger = require( '@stdlib/math/base/assert/is-nonnegative-integer' );
```
```suggestion
var isNonNegativeIntegerf = require( '@stdlib/math/base/assert/is-nonnegative-integerf' );
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/bernoullif/lib/main.js:
```diff
@@ -0,0 +1,104 @@
+		return 0.0;
+	}
+	if ( n > MAX_BERNOULLI ) {
+		return ( ( n / 2 ) & 1 ) ? PINF : NINF;
```
```suggestion
		return ( (n/2)&1 ) ? PINF : NINF;
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/bernoullif/examples/index.js:
```diff
@@ -0,0 +1,29 @@
+*/
+
+'use strict';
+
+var bernoullif = require( './../lib' );
+
+var v;
+var i;
+
+for ( i = 0; i < 70; i++ ) {
+	v = bernoullif( i );
+	console.log( v );
+}
```
```suggestion
var logEachMap = require( '@stdlib/console/log-each-map' );
var discreteUniform = require( '@stdlib/random/array/discrete-uniform' );
var bernoullif = require( './../lib' );

var x = discreteUniform( 100, 0, 70, {
	'dtype': 'int32'
});

logEachMap( 'bernoullif(%d) = %0.4f', x, bernoullif );
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/bernoullif/examples/c/example.c:
```diff
@@ -0,0 +1,31 @@
+
+	for ( i = 0; i < 70; i++ ) {
+		v = stdlib_base_bernoullif( i );
+		printf( "bernoulli(%d) = %f\n", i, v );
```
```suggestion
		printf( "bernoullif(%d) = %f\n", i, v );
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/bernoullif/test/test.js:
```diff
@@ -0,0 +1,109 @@
+
+	for ( i = -1; i > -50; i-- ) {
+		v = bernoullif( i );
+		t.strictEqual( isnanf( v ), true, 'returns expected value when provided ' + i );
```
```suggestion
		t.strictEqual( isnanf( v ), true, 'returns expected value' );
```

Applies here and below...

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/bernoullif/test/test.js:
```diff
@@ -0,0 +1,109 @@
+tape( 'the function returns the nth Bernoulli number for odd numbers', function test( t ) {
+	var v;
+	var i;
+	for ( i = 3; i < 500; i += 2 ) {
```
```suggestion
	for ( i = 3; i < 100; i += 2 ) {
```

A smaller range is ok for `bernoullif`

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/bernoullif/test/test.js:
```diff
@@ -0,0 +1,109 @@
+tape( 'the function returns +/- infinity for large integers', function test( t ) {
+	var v;
+	var i;
+	for ( i = 66; i < 500; i += 2 ) {
```
```suggestion
	for ( i = 66; i < 150; i += 2 ) {
```
Same comment.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/bernoullif/benchmark/benchmark.js:
```diff
@@ -0,0 +1,52 @@
+// MODULES //
+
+var bench = require( '@stdlib/bench' );
+var randu = require( '@stdlib/random/array/discrete-uniform' );
```
```suggestion
var discreteUniform = require( '@stdlib/random/array/discrete-uniform' );
```

For being consistent across other packages.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/bernoullif/benchmark/benchmark.js:
```diff
@@ -0,0 +1,52 @@
+	var y;
+	var i;
+
+	x = randu( 100, 0, 500 );
```
```suggestion
	x = discreteUniform( 100, 0, 100 );
```

100 is fine here for the max value.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/bernoullif/benchmark/c/native/benchmark.c:
```diff
@@ -0,0 +1,136 @@
+	int i;
+
+	for ( i = 0; i < 100; i++ ) {
+		x[ i ] = ( 500.0 * rand_float() );
```
```suggestion
		x[ i ] = ( 500.0f * rand_float() );
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/bernoullif/benchmark/c/native/benchmark.c:
```diff
@@ -0,0 +1,136 @@
+
+	t = tic();
+	for ( i = 0; i < ITERATIONS; i++ ) {
+		y = stdlib_base_bernoullif( (int)( x[ i % 100 ] ) );
```
```suggestion
		y = stdlib_base_bernoullif( (int32_t)( x[ i%100 ] ) );
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/bernoullif/manifest.json:
```diff
@@ -0,0 +1,78 @@
+      "libraries": [],
+      "libpath": [],
+      "dependencies": [
+        "@stdlib/math/base/assert/is-odd",
```
`@stdlib/math/base/assert/is-odd` is not needed here now.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/bernoullif/package.json:
```diff
@@ -0,0 +1,66 @@
+    "stdmath",
+    "mathematics",
+    "math",
+    "special functions",
```
```suggestion
```
Keywords with more than one word are not used.

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/bernoullif/README.md:
````diff
@@ -0,0 +1,232 @@
+<!-- eslint no-undef: "error" -->
+
+```javascript
+var bernoullif = require( '@stdlib/math/base/special/bernoullif' );
+
+var v;
+var i;
+
+for ( i = 0; i < 70; i++ ) {
+    v = bernoullif( i );
+    console.log( v );
+}
````
```suggestion
var logEachMap = require( '@stdlib/console/log-each-map' );
var discreteUniform = require( '@stdlib/random/array/discrete-uniform' );
var bernoullif = require( './../lib' );

var x = discreteUniform( 100, 0, 70, {
    'dtype': 'int32'
});

logEachMap( 'bernoullif(%d) = %0.4f', x, bernoullif );
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/bernoullif/README.md:
````diff
@@ -0,0 +1,232 @@
+
+```c
+float out = stdlib_base_bernoullif( 0 );
+// returns 1.0
````
```suggestion
// returns 1.0f
```

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/bernoullif/README.md:
```diff
@@ -0,0 +1,232 @@
+// returns 1.0
+
+out = stdlib_base_bernoullif( 1 );
+// returns 0.5
```
```suggestion
// returns 0.5f
```

anandkaranubc committed:
chore: apply suggestions from review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: passed
  - task: lint_c_examples
    status: passed
  - task: lint_c_benchmarks
    status: passed
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter commented on lib/node_modules/@stdlib/math/base/special/bernoullif/README.md:
```diff
@@ -0,0 +1,232 @@
+
+#### stdlib_base_bernoullif( n )
+
+Compute the nth [Bernoulli number][bernoulli-number] as a single-precision floating-point number.
```
```suggestion
Computes the nth [Bernoulli number][bernoulli-number] as a single-precision floating-point number.
```

Planeshifter committed:
docs: update description

Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>
